### PR TITLE
feat: add a popup showing "Updating firmware"

### LINKED ISSUE #251: Display "Updating firmware" when the erase-and-reset command is received
### Description

When PanelDue received the erase-and-reset command from RRF, the screen freezes. It would be better if it cleared the screen and displayed "Updating firmware" before executing the erase-and-reset command.

### PR DESCRIPTION
**Pull request by wilriker**
Closes #251

### CODE REVIEW

File: src/UI/UserInterface.cpp [hunk 1]
```diff
@@ -1158,6 +1167,7 @@ static void CreateMainPages(uint32_t language, const ColourScheme& colours)
 	CreateMessageTabFields(colours);
 	CreateSetupTabFields(language, colours);
 	CreateScreensaverPopup();
+	CreateFirmwareUpdatePopup();
```
Comment (mfs12) [inline src/UI/UserInterface.cpp:1170]: Please move call to CreateFields where all other Popups are created.

File: src/UI/UserInterface.cpp [hunk 2]
```diff
@@ -685,6 +685,15 @@ static void CreateScreensaverPopup()
 	screensaverPopup->AddField(screensaverText = new StaticTextField(row1, margin, screensaverTextWidth, TextAlignment::Left, text));
 }
 
+static void CreateFirmwareUpdatePopup()
+{
+	firmwareUpdatePopup = new PopupWindow(max(DisplayX, DisplayY), max(DisplayX, DisplayY), black, black, false);
+	DisplayField::SetDefaultColours(white, black);
+	static const char * text = "Updating firmware";
```
Comment (mfs12) [inline src/UI/UserInterface.cpp:692]: please fix formatting.

Comment (mfs12) [reply src/UI/UserInterface.cpp:692]: Also string should be used from strings.hpp.

**Comment by mfs12** [comment]
Hey @wilriker, if I understand @dc42 correctly

> When PanelDue receives the erase-and-reset command from RRF, the screen freezes. It would be better if it cleared the screen and displayed "Updating firmware" before executing the erase-and-reset command.

The idea is not to have a popup, but a full main page covering everything.

Also I think if it should be "just" a popup RRF could just send a message which then will displayed as well. This would mean no change PanelDue side.

@dc42 can comment on this idea.

**Comment by wilriker** [comment]
> The idea is not to have a popup, but a full main page covering everything.

That's exactly what this PR achieves. The popup covers the full screen in black and positions the text in white at the center. Doing this as a logical popup instead of low-level overwriting the even in black and then painting the text simply reuses the existing UI framework making all the required code much shorter. But it's the exact same approach that is used by the screensaver.

> Also I think if it should be "just" a popup RRF could just send a message which then will displayed as well. This would mean no change PanelDue side.

I'm not certain if that would work over a static popup. Reason is that for flashing from Duet we are on a tight time budget to have the Duet recognize that the PanelDue in fact did erase and reset. In my first code I preventatively added a 100ms delay to have the screen repaint but that broke flashing completely (and it turned out it was not necessary anyways).

**Comment by dc42** [comment]
I don't mind how this is implemented, and I can quite understand that clearing the screen and then using the existing popup code may be the simplest way to achieve this. I would rather not have to add code to RRF to do this because we are already running out of flash memory space on Duet 2.

**Comment by mfs12** [comment]
After discussion on slack, @dc42 confirmed this solution.

**Pull request merged**